src/qibo/derivative.py

```diff
@@ -106,19 +118,49 @@ def circuit(nqubits = 1):
     shifted[parameter_index] += s
     circuit.set_parameters(shifted)
 
-    forward = hamiltonian.expectation(
-        backend.execute_circuit(circuit=circuit, initial_state=initial_state).state()
-    )
+    forward = 0
+    backward = 0
 
-    # backward shift and evaluation
-    shifted[parameter_index] -= 2 * s
-    circuit.set_parameters(shifted)
+    if nshots == None:
+        forward = hamiltonian.expectation(
+            backend.execute_circuit(
+                circuit=circuit, initial_state=initial_state
+            ).state()
+        )
+
+        # backward shift and evaluation
+        shifted[parameter_index] -= 2 * s
+        circuit.set_parameters(shifted)
+
+        backward = hamiltonian.expectation(
+            backend.execute_circuit(
+                circuit=circuit, initial_state=initial_state
+            ).state()
+        )
+
+    else:
+        copied = shifted.copy()
```

Comment:
Is there any reason you need an additional copy of the parameters when `nshots` are given? In principle you can use the same approach as before, having a single copy on which you do the shifts.